lib/ProductOpener/FoodGroups.pm

```diff
@@ -0,0 +1,437 @@
+				$log->debug("found food group for category", { category_id => $categoryid, food_groups => $product_ref->{food_groups} }) if $log->is_debug();
+                last;
+            }
+        }
+    }
+    
+	# Compute the food groups tags, including parents, in food_groups_tags
+	$product_ref->{food_groups_tags} = [ gen_tags_hierarchy_taxonomy("en", "food_groups", $product_ref->{food_groups}) ];
+
+	# Compute old PNNS groups tags, for comparison.
+	# Will eventually be removed.
+    compute_pnns_groups($product_ref);
+
+	# Put back the original categories_tags so that they match what is in the taxonomy field
+	# if there is a mistmatch it can cause tags to be added multiple times (e.g. with imports)
+	if (defined $product_ref->{original_categories_tags}) {
+		$product_ref->{categories_tags} = [@{$product_ref->{original_categories_tags}}];
+		delete $product_ref->{original_categories_tags};
+	}	
+}
+
+
+=head2 temporarily_change_categories_for_food_groups_computation ( $product_ref )
+
+Food groups are derived from categories.
+In order to account to some subtleties (categories in OFF are manually entered and may have a broad and imprecise scope:
+e.g. what is a beverage?), this function has some rules to temporarily change categories, so that:
+
+- only products that matche the precise definition of a beverage according to the Nutri-Score formula are counted as beverages
+(e.g. a beverage with more than 80% milk is not counted as beverage)
+- alcoholic beverages with less than 1% alcohol are not counted as alcoholic beverages
+- beverages with sweeteners / added sugar are automatically categorized as artificially sweetened / sweetened beverages
+
+=head3 Arguments
+
+=head4 product reference $product_ref
+
+=head3 Return values
+
+Categories can be added or removed in $product_ref->{categories_tags}
+
+Original categories are saved in $product_ref->{original_categories_tags}
+
+=cut
+
+sub temporarily_change_categories_for_food_groups_computation($) {
+
+	my $product_ref = shift;
+
+	# Only add or remove categories tags temporarily for determining the food groups / PNNS groups
+	# save the original value
+
+	if (defined $product_ref->{categories_tags}) {
+		$product_ref->{original_categories_tags} = [@{$product_ref->{categories_tags}}];
+	}
+
+	# For Open Food Facts, add special categories for beverages that are computed from
+	# nutrition facts (alcoholic or not) or the ingredients (sweetened, artificially sweetened or unsweetened)
+	# those tags are only added to categories_tags (searchable in Mongo)
+	# and not to the categories_hierarchy (displayed on the web site and in the product edit form)
+	# Those extra categories are also used to determined the French PNNS food groups
+
+	if (($product_ref->{nutrition_score_beverage}) and (not has_tag($product_ref,"categories","en:instant-beverages"))) {
+
+		if (defined $product_ref->{nutriments}{"alcohol_100g"}) {
```

Comment:
suggestion:
I would have made two arrays: to_remove and to_add, to collect categories to eventually add / remove.
And only at the end I would have made a loop to remove / add what was to remove / add.
I think, It would make the code more easy to read.